tx-parser/src/lib.rs

```diff
@@ -0,0 +1,378 @@
+// specific language governing permissions and limitations under the License.
+
+#![allow(dead_code)]
+
+extern crate edn;
+extern crate combine;
+extern crate mentat_tx;
+
+use combine::{any, eof, many, optional, parser, satisfy_map, token, Parser, ParseResult, Stream};
+use combine::combinator::{Expected, FnParser};
+// TODO: understand why this is self::edn rather than just edn.
+use self::edn::types::Value;
+use mentat_tx::entities::*;
+
+// TODO: implement combine::Positioner on Value.  We can't do this
+// right now because Value is defined in edn and the trait is defined
+// in combine.
+
+// #[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+// pub struct ValuePosition {
+//     pub position: usize,
+// }
+
+// impl fmt::Display for ValuePosition {
+//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+//         write!(f, "value position: {}", self.position)
+//     }
+// }
+
+// impl combine::primitives::Positioner for Value {
+//     type Position = ValuePosition;
+//     fn start() -> ValuePosition {
+//         ValuePosition { position: 1 }
+//     }
+//     fn update(&self, position: &mut ValuePosition) {
+//         position.position += 1;
+//     }
+// }
+
+struct Tx<I>(::std::marker::PhantomData<fn(I) -> I>);
+
+type TxParser<O, I> = Expected<FnParser<I, fn(I) -> ParseResult<O, I>>>;
+
+fn fn_parser<O, I>(f: fn(I) -> ParseResult<O, I>, err: &'static str) -> TxParser<O, I>
+    where I: Stream<Item = Value>
+{
+    parser(f).expected(err)
+}
+
+impl<I> Tx<I>
+    where I: Stream<Item = Value>
+{
+    fn integer() -> TxParser<i64, I> {
+        fn_parser(Tx::<I>::integer_, "integer")
+    }
+
+    fn integer_(input: I) -> ParseResult<i64, I> {
+        return satisfy_map(|x: Value| if let Value::Integer(y) = x {
+                Some(y)
+            } else {
+                None
+            })
+            .parse_stream(input);
+    }
+
+    fn keyword() -> TxParser<String, I> {
+        fn_parser(Tx::<I>::keyword_, "keyword")
+    }
+
+    fn keyword_(input: I) -> ParseResult<String, I> {
+        return satisfy_map(|x: Value| if let Value::Keyword(y) = x {
+                Some(y)
+            } else {
+                None
+            })
+            .parse_stream(input);
+    }
+
+    fn entid() -> TxParser<EntId, I> {
+        fn_parser(Tx::<I>::entid_, "entid")
+    }
+
+    fn entid_(input: I) -> ParseResult<EntId, I> {
+        let p = Tx::<I>::integer()
+            .map(|x| EntId::EntId(x))
+            .or(Tx::<I>::keyword().map(|x| EntId::Ident(x)))
+            .parse_lazy(input)
+            .into();
+        return p;
+    }
+
+    fn lookup_ref() -> TxParser<LookupRef, I> {
+        fn_parser(Tx::<I>::lookup_ref_, "lookup-ref")
+    }
+
+    fn lookup_ref_(input: I) -> ParseResult<LookupRef, I> {
+        return satisfy_map(|x: Value| if let Value::Vector(y) = x {
+                let mut p = (Tx::<&[Value]>::entid(), any(), eof())
```

Comment:
No, and there's a subtle thing happening with consuming `Parser` instances that I don't yet understand.  You'll note all these parsers are mutable; that's because they're consumed by `parse` and friends.  So you _can't_ avoid the invocations and allocations.  It's unclear how this impacts performance overall.